### ISSUE TITLE
Seeing bizarre errors all over ATT, extensions broken despite being enabled.

### ISSUE THREAD

**Issue by Arcane21** [Orain/ansible-playbook#650]
Over half our site features are broken despite being enabled, more details here:

http://drunkardswalkforums.yuku.com/reply/130990/All-The-Tropes-Wiki-Project-Part-V#reply-130990

My guess is something broke for ATT when the farms were completely restored, giving ATT another problem, since I not sure what, I'd like to ask someone else too look into it, since I don't see any obvious issues.

Ansible update and LC refresh do not fix the issue.

**Comment by Arcane21** [comment]
Issues still persist, have no clue why since they initially started to clear up when I LC refreshed all MW servers

**Comment by Arcane21** [comment]
I'm having to run a LC refresh at least once a day to keep things from breaking, heavily suspect it might have something to do with the Translate extension, which has had a recent reported issue with breaking certain extensions according to Phabricator.

**Comment by Arcane21** [comment]
Rebuilding ar...
Rebuilding ca...
Rebuilding da...
Rebuilding de...
Rebuilding el...
Rebuilding en-gb...
Rebuilding es...
Rebuilding fa...
Rebuilding fr...
Rebuilding he...
Rebuilding id...
Rebuilding it...
Rebuilding ja...
Rebuilding ko...
Rebuilding mk...
Rebuilding nb...
Rebuilding nl...
Rebuilding pl...
Rebuilding pt...
Rebuilding pt-br...
Rebuilding ru...
Rebuilding se...
Rebuilding si...
Rebuilding sk...
Rebuilding sv...
Rebuilding tr...
Rebuilding zh...
Rebuilding zh-tw...

These same languages consistently have to be rebuilt.

I again reiterate a fairly safe patch until a more permanent solution can be found would be an autorun LC refresh on all MW servers until the root cause can be sufficiently addressed.

**Comment by NDKilla** [comment]
rebuildLocalisationCache.php --wiki extloadwiki --force?

Wonder how all the messages on ATT look after rebuilding en (not en-gb) and others

**Comment by Arcane21** [comment]
Same basic result, more or less.

**Comment by NDKilla** [comment]
Arcane if you're still around and can shell right now can you enable extensions/Spoilers/Spoilers.php on extloadwiki and force rebuild the cache. I don't wanna do anything when I'm not around to test / make sure it works. If you're busy don't worry bout it

**Comment by addshore** [comment]
We need to determine what is falling out of the LC cache? / never making it into the LC cache / why.
Again looking from afar this looks like something in ATT is not on extload, thus messages and magic words are missing.

**Comment by Arcane21** [comment]
Checked, all exts that could possibly be an issue are on extload.

Also have tested the timing, seems the LC degenerates every 18-24 hours, so having a silent auto LC refresh run at least twice a day would be sufficient to hold this error well in abeyance.

**Comment by Arcane21** [comment]
Has any progress been made on getting an autorun job set up? I'm still having to do it manually at least twice a day.

**Comment by Arcane21** [comment]
Saw an interesting error last few time I rebuilt the LC cache:


[40da3ed2] [no req]   Exception from line 559 of /srv/mediawiki/w/includes/cache/LocalisationCache.php: LocalisationCache::readJSONFile: Invalid JSON file: /srv/mediawiki/w/extensions/CreateWiki/i18n/fr.json
Backtrace:
#0 /srv/mediawiki/w/includes/cache/LocalisationCache.php(885): LocalisationCache->readJSONFile(string)
#1 /srv/mediawiki/w/includes/cache/LocalisationCache.php(1435): LocalisationCache->recache(string)
#2 /srv/mediawiki/w/maintenance/rebuildLocalisationCache.php(162): LocalisationCacheBulkLoad->recache(string)
#3 /srv/mediawiki/w/maintenance/rebuildLocalisationCache.php(129): RebuildLocalisationCache->doRebuild(array, LocalisationCacheBulkLoad, boolean)
#4 /srv/mediawiki/w/maintenance/doMaintenance.php(101): RebuildLocalisationCache->execute()
#5 /srv/mediawiki/w/maintenance/rebuildLocalisationCache.php(181): require_once(string)
#6 {main}

**Comment by NDKilla** [comment]
Is create wiki breaking LC? Hmm

**Comment by Reception123** [comment]
@NDKilla  I think so. From the first day I started creating wikis, after creating them I always get a blank page, at first I thought it was normal.

**Comment by addshore** [comment]
This CreateWiki issue has been fixed in https://github.com/Orain/CreateWiki/commit/066147c59246a34394fbab2ec7c63c11b329e42f

I do not believe this is the cause of the overall issue

**Comment by Arcane21** [comment]
Confirmed the cache breaks down on all wikis at the same time. Setting a forced refresh job on all servers at least once every hour will definitely keep this issue in abeyance indefinitely.

**Comment by Arcane21** [comment]
Okay, made another observation while observing wikis, we'll need to do a forced LC and database update, this will ensure the cache flushes any database specific errors as well.

Speaking of that, saw some CentralNotice errors on some wikis during the database update job I ran, someone might want to look into that.

**Comment by Arcane21** [comment]
Okay, nailed down everything that goes wrong and my fixes for it.

The localization cache and lua seem to both degenerate after a few hours each day due to certain extensions with known database/caching issues as the source of the problem, and since we can't remove the extensions due to them being crucial, we need to set up a forced LC Refresh/ansible update at least once and hours on all MW servers simultaneously.

This will not only fix these issues, but provide a nice sanity check for ansible updates in general, insuring we remain constantly updated.

**Comment by Arcane21** [comment]
I just changed LocalisationUpdate and Babel to the master branch in our repo, both have gone through fine without incident, will need to see if we have another LC degeneration event to be sure, but I have verified making these changes has not hurt anything.

According to Phabricator and the recent commits on the WMF repo, this should fix several recurring errors and hopefully obviate the need to continually have to rerun LC refreshes, and if this does fix our issues, I planned to changed several other extensions to use the master branch since I saw a few have had a lot of fixes of other issues I've seen reported on Orain.

**Comment by addshore** [comment]
=o any news on the above and if it fixed 'everything'? :D

**Comment by Arcane21** [comment]
Didn't do squat. We need to upgrade a lot more, including MW core.

**Comment by NDKilla** [comment]
@Nikerabbit as mentioned in the bug report, our config is available [here](https://github.com/Orain/ansible-playbook/blob/master/roles/mediawiki/files/localsettings/LocalSettings.php.j2). Ctrl+F the page for cache and you can see that setting, including several others. The information in this file is what is used on our servers so, you can just find it.

But since you asked for a specific example, 
```php
// Don't care about LC cache in maintenance scripts, otherwise it
// might throw errors (e.g. during a server setup where rebuildLocalisationCache.php
// automatically runs, and it can't find a cache, so
// it throws an exception and then dies).

if ( !$wgCommandLineMode ) { 
	$wgLocalisationCacheConf['manualRecache'] = true;
}
```

**Comment by Nikerabbit** [comment]
So it seems you have `$wgCacheDirectory` which means that you have LC cache stored in files, same location for each wiki.

This raises few questions:
* is anything deleting files from that directory (it's under `/tmp`) periodically?
* if you run different wikis having different set of extensions, the cache would be constantly regenerated. Do you have a workaround for that?

Some messages, especially those for JavaScript are further cached separately in the DB. If a cache happens to be missing during a maintenance script (when you prevent recaching), it might be cached with missing messages.

I suggest watching timestamps on those cache files to see if they are constantly updating.

**Comment by NDKilla** [comment]
Thanks for the ideas.. I'm not really sure and I'll mention your comment to @addshore, Currently every single extension downloaded and enabled anywhere is loaded on "extloadwiki " a private wiki we use for this purpose. When we rebuild the LC cache we force rebuild it on that wiki only, and I believe other wikis might use the languages there as needed... I'm not exactly sure how it all works tho :)

**Comment by addshore** [comment]
So yes, we have a single wiki with every loaded, we generate the cache files for that wiki and those files are then used for all other wikis. Hence the rule of everything must be loaded on the extloadwiki.

Right now I don't have any time to look at the timestamp files etc but I am sure @NDKilla can keep an eye on them.

We have nothing setup, that I am aware of, flushing our tmp dir

**Comment by NDKilla** [comment]
Merged #720 and issue seems fixed, on all wikis. 
I merged the above PR, manually ran ansible on all wikis, force rebuild extloadwiki, and refreshed pages and it appears to work.
This may be for now because I just rebuilt it, but based on the fact that all the rebuilds took longer than usual, I'm hoping it did something different.
Closing this as done now with huge thanks to @Southparkfan and hoping it doesn't get re-opened.
